microtiter_plates/libraries/microtiterplates/source.rb

```diff
@@ -164,10 +198,73 @@ def associate_group(group:, key:, data:)
     group.each { |i| associate(index: i, key: key, data: data) }
   end
 
-  private
+  # Uses `PartProvenance` to associate the provided provenance data to
+  #   indices of the provided group
+  #
+  # @param group [Array<Array<Fixnum>>]
+  # @param key [String] the key pointing to the relevant `DataAssociation`
+  # @param data [Array<Hash>] the data for the association; each Hash must
+  #   include an `item: Item` pair
+  # @return [void]
+  def associate_provenance_group(group:, key:, data:)
+    group.each { |i| associate_provenance(index: i, key: key, data: data) }
+  end
 
+  # Make a simple data association on a part
+  #
+  # @param index [Array<Fixnum>] the row, column pair pointing to the part
+  # @param key [String] the key pointing to the relevant `DataAssociation`
+  # @param data [serializable object]  the data for the association
+  # @return [void]
   def associate(index:, key:, data:)
     part = @collection.part(index[0], index[1])
     part.associate(key, data)
   end
+
+  # Uses `PartProvenance` to associate the provided provenance data to
+  #   a part
+  #
+  # @param index [Array<Fixnum>] the row, column pair pointing to the part
+  # @param key [String] the key pointing to the relevant `DataAssociation`
+  # @param data [serializable object] the data for the association; Hash must
+  #   include an `item: Item` pair
+  # @return [void]
+  def associate_provenance(index:, key:, data:)
```

Comment:
This method does not seem to be working well for assigning destination provenance to the incoming items. Rather than updating a single `destination` association with additional parts, it keeps adding successively longer arrays under the same key and the key is simply duplicated. @bjkeller

Reply:
![image](https://user-images.githubusercontent.com/7586251/87860828-75f8be00-c8f5-11ea-8f73-6c1737431093.png)
![image](https://user-images.githubusercontent.com/7586251/87860842-93c62300-c8f5-11ea-90c5-0e61977c20c6.png)
![image](https://user-images.githubusercontent.com/7586251/87860858-ae000100-c8f5-11ea-9375-39d35871ae6a.png)

Reply:
It also seems to be duplicating other keys, here `volume`

Reply:
```ruby
# typed: false
# frozen_string_literal: true

needs 'Standard Libs/AssociationManagement'
needs 'Standard Libs/Debug'

class Protocol

  include AssociationManagement
  include PartProvenance
  include Debug

  def main

    operations.make

    input_item = operations.first.input('Input').item
    input_item.associate(:foo, 'bar')

    output_items = [
      operations.first.output('Output 1').item,
      operations.first.output('Output 2').item,
      operations.first.output('Output 3').item
    ]

    output_items.zip(['bax', 'bay', 'baz']).each do |a, b|
      a.associate(:foo, b)
    end

    output_items.each do |output_item|
      add_one_to_one_provenance(from_item: input_item, to_item: output_item)
      inspect input_item.associations, 'input'
      inspect output_item.associations, 'output'
    end

    {}

  end

  # Add provenance data to a source-destination pair of items
  #
  # @param from_item [Item]
  # @param to_item [Item]
  # @param additional_relation_data [serializable object] additional data that
  #   will be added to the provenace association
  # @return [void]
  def add_one_to_one_provenance(from_item:, to_item:,
                                additional_relation_data: nil)
    from_map = AssociationMap.new(from_item)
    to_map = AssociationMap.new(to_item)

    add_provenance(
      from: from_item, from_map: from_map,
      to: to_item, to_map: to_map,
      additional_relation_data: additional_relation_data
    )
    from_map.save
    to_map.save
  end

end
```

produces:

```
**2020-07-19T17:28:56.066-07:00**
- {"title":"<span style=\"background-color:yellow\">INSPECTING input (ActiveSupport::HashWithIndifferentAccess)</span>"}
- {"note":"{\"foo\":\"bar\",\"destination\":[{\"id\":12602}]}"}
**2020-07-19T17:28:56.200-07:00**
- {"title":"<span style=\"background-color:yellow\">INSPECTING output (ActiveSupport::HashWithIndifferentAccess)</span>"}
- {"note":"{\"foo\":\"bax\",\"source\":[{\"id\":12601}]}"}
**2020-07-19T17:28:56.204-07:00**
- {"title":"<span style=\"background-color:yellow\">INSPECTING input (ActiveSupport::HashWithIndifferentAccess)</span>"}
- {"note":"{\"foo\":\"bar\",\"destination\":[{\"id\":12602},{\"id\":12603}]}"}
**2020-07-19T17:28:56.221-07:00**
- {"title":"<span style=\"background-color:yellow\">INSPECTING output (ActiveSupport::HashWithIndifferentAccess)</span>"}
- {"note":"{\"foo\":\"bay\",\"source\":[{\"id\":12601}]}"}
**2020-07-19T17:28:56.225-07:00**
- {"title":"<span style=\"background-color:yellow\">INSPECTING input (ActiveSupport::HashWithIndifferentAccess)</span>"}
- {"note":"{\"foo\":\"bar\",\"destination\":[{\"id\":12602},{\"id\":12603},{\"id\":12604}]}"}
**2020-07-19T17:28:56.251-07:00**
- {"title":"<span style=\"background-color:yellow\">INSPECTING output (ActiveSupport::HashWithIndifferentAccess)</span>"}
- {"note":"{\"foo\":\"baz\",\"source\":[{\"id\":12601}]}"}
```

Reply:
However, when you run the operation through the manager, the input item looks like this:
![image](https://user-images.githubusercontent.com/7586251/87889536-f814df80-c9e6-11ea-8836-e587789aace2.png)

Reply:
So it looks to me like:

1. The `add_provenance` method is adding new rows to the association table, rather than overwriting them. 
2. The `item.associations` method returns only the last one for any given key
3. The GUI shows all of them

Reply:
@bjkeller what should we do here?